zkevm-circuits/src/state_circuit/state.rs

```diff
@@ -605,379 +456,119 @@ impl<
-        storage_key: F,
-        value_prev: F,
-    ) -> Result<BusMapping<F>, Error> {
-        let address = {
-            let cell = region.assign_advice(|| "address", self.address, offset, || Ok(address))?;
-            Variable::<F, F>::new(cell, Some(address))
-        };
+        row: RwRow<F>,
+        row_prev: RwRow<F>,
+        diff_is_zero_chips: &[IsZeroChip<F>; 5],
+    ) -> Result<(), Error> {
+        let address = row.key3;
+        let rw_counter = row.rw_counter;
+        let value = row.value;
+        let is_write = row.is_write;
 
-        if SANITY_CHECK && rw_counter > F::from(RW_COUNTER_MAX as u64) {
-            panic!("rw_counter out of range");
+        // check witness sanity
+        if offset > ROWS_MAX {
+            panic!("too many storage operations");
         }
-        let rw_counter = {
-            let cell = region.assign_advice(
-                || "global counter",
-                self.rw_counter,
-                offset,
-                || Ok(rw_counter),
-            )?;
-
-            Variable::<F, F>::new(cell, Some(rw_counter))
-        };
-
-        let value = {
-            let cell = region.assign_advice(|| "value", self.value, offset, || Ok(value))?;
-
-            Variable::<F, F>::new(cell, Some(value))
-        };
-
-        let storage_key = {
-            let cell = region.assign_advice(
-                || "storage key",
-                self.storage_key,
-                offset,
-                || Ok(storage_key),
-            )?;
-
-            Variable::<F, F>::new(cell, Some(storage_key))
-        };
-
-        let value_prev = {
-            let cell = region.assign_advice(
-                || "value prev",
-                self.value_prev,
+        if SANITY_CHECK {
+            if rw_counter > F::from(RW_COUNTER_MAX as u64) {
+                panic!("rw_counter out of range");
+            }
+            if row.tag == F::from(STACK_TAG as u64) && address > F::from(STACK_ADDRESS_MAX as u64) {
+                panic!(
+                    "stack address out of range {:?} > {}",
+                    address, STACK_ADDRESS_MAX
+                );
+            }
+            if row.tag == F::from(MEMORY_TAG as u64) && address > F::from(MEMORY_ADDRESS_MAX as u64)
+            {
+                panic!(
+                    "memory address out of range {:?} > {}",
+                    address, MEMORY_ADDRESS_MAX
+                );
+            }
+        }
+        region.assign_fixed(|| "enable row", self.s_enable, offset, || Ok(F::one()))?;
+        region.assign_advice(|| "rw counter", self.rw_counter, offset, || Ok(rw_counter))?;
+        region.assign_advice(|| "value", self.value, offset, || Ok(value))?;
+        region.assign_advice(|| "is_write", self.is_write, offset, || Ok(is_write))?;
+
+        for (i, diff_is_zero_chip) in diff_is_zero_chips.iter().enumerate() {
+            let (value, diff) = match i {
+                // FIXME: find a better way here
```

Comment:
I think this is good enough :)
The only think that would come to my mind to improve this would be to change `RwRow` to have a field `keys: [F, 5]`.